### PR TITLE
workers: basic event loop

### PR DESCRIPTION
**Pull request by bartlomieju**
* establish basic event loop for workers
* make "self.close()" inside worker
* remove "runWorkerMessageLoop() - instead manually call global function in Rust when message arrives. This is done in preparation for structured clone
* refactor "WorkerChannel" and use distinct structs for internal and external channels; "WorkerChannelsInternal" and "WorkerHandle"
* move "State.worker_channels_internal" to "Worker.internal_channels"
* add "WorkerEvent" enum for child->host communication; currently "Message(Buf)" and "Error(ErrBox)" variants are supported
* add tests for nested workers
* add tests for worker throwing error on startup

### CODE REVIEW

File: cli/ops/worker_host.rs [hunk 1]
```diff
@@ -37,6 +44,190 @@ pub fn init(i: &mut Isolate, s: &ThreadSafeState) {
...
+    let result = if has_source_code {
+      worker.execute(&source_code)
+    } else {
+      // TODO(bartlomieju): add "type": "classic", ie. ability to load
+      // script instead of module
+      let load_future = worker
+        .execute_mod_async(&specifier, None, false)
+        .boxed_local();
+
+      rt.block_on(load_future)
```
Comment (bartlomieju) [inline cli/ops/worker_host.rs]: This should be `select!()` between module execution and termination - https://html.spec.whatwg.org/multipage/workers.html#worker-processing-model - point 24

File: cli/ops/worker_host.rs [hunk 2]
```diff
@@ -37,6 +44,190 @@ pub fn init(i: &mut Isolate, s: &ThreadSafeState) {
...
+    {
+      let mut_channels = worker.state.worker_channels_internal.lock().unwrap();
+      let channels = mut_channels.as_ref().unwrap().clone();
+      let msg = serialize_worker_result(result)
+        .to_string()
+        .into_boxed_str()
+        .into_boxed_bytes();
+      futures::executor::block_on(channels.post_message(msg))
+        .expect("Failed to post message to host");
+    }
```
Comment (bartlomieju) [inline cli/ops/worker_host.rs]: This bit should be called only on error during module loading and this block should return

File: cli/ops/worker_host.rs [hunk 3]
```diff
@@ -37,6 +44,190 @@ pub fn init(i: &mut Isolate, s: &ThreadSafeState) {
...
+          let channels = mut_channels.as_ref().unwrap().clone();
+          channels.get_message()
+        };
+
+        let _result =
+          match futures::future::select(&mut *worker, receive_msg_fut).await {
+            Either::Left((worker_result, _msg_fut)) => match worker_result {
+              Ok(()) => {
+                // worker finished scripts, no point in polling it
+                // until we receive next message from host (not valid if we add unref
```
Comment (bartlomieju) [inline cli/ops/worker_host.rs]: Probably https://docs.rs/futures/0.3.4/futures/future/fn.select_all.html between:
- receive message from host
- receive termination from host
- receive close from worker
- if worker has pending ops: then worker future

File: cli/js/workers_test.ts [hunk 4]
```diff
@@ -0,0 +1,87 @@
+// Copyright 2018-2020 the Deno authors. All rights reserved. MIT license.
+import { test, assert, assertEquals } from "./test_util.ts";
+import { runIfMain } from "../../std/testing/mod.ts";
```
Comment (ry) [inline cli/js/workers_test.ts]: I don't think we should import from std ... afaik this is not done in cli/js.. You can get runIfMain from cli/js/test_util.ts

File: foo.ts [hunk 5]
```diff
@@ -0,0 +1,3 @@
+import * as lodash from "https://cdn.jsdelivr.net/npm/lodash@4.17.15/lodash.min.js";
+
+console.log(lodash);
```
Comment (ry) [inline foo.ts]: Remove

Comment (bartlomieju) [reply foo.ts]: Oooops

**Comment by bartlomieju** [comment]
> Doesn't seem to enable 026_workers ?

Replaced with unit tests in `cli/js/workers_test.ts`. Removed `026_workers` from integration tests

File: cli/web_worker.rs [hunk 6]
```diff
@@ -125,29 +131,29 @@ mod tests {
...
     });
 
-    let r = tokio_util::run_basic(worker);
+    let mut rt = tokio_util::create_basic_runtime();
+    let r = run_worker_loop(&mut rt, &mut worker);
     assert!(r.is_ok())
   }
 
+  // FIXME(bartlomieju)
+  #[ignore]
```
Comment (ry) [inline cli/web_worker.rs]: Are the changes to this test necessary? I'm fine with it getting disabled but maybe it shouldn't also change behavior.

Comment (bartlomieju) [reply cli/web_worker.rs]: Yes; `onmessage = () => { delete self.onmessage; }` makes no sense. I'll try to fix those tests now

File: cli/ops/worker_host.rs [hunk 7]
```diff
@@ -34,7 +42,159 @@ pub fn init(i: &mut Isolate, s: &State) {
...
+    WebWorker::new(name.to_string(), startup_data::deno_isolate_init(), state);
+  let script = format!("bootstrapWorkerRuntime(\"{}\")", name);
+  worker.execute(&script)?;
+
+  Ok(worker)
+}
+
+// TODO(bartlomieju): this function should probably live in `cli/web_worker.rs`
+pub fn run_worker_loop(
+  rt: &mut tokio::runtime::Runtime,
```
Comment (ry) [inline cli/ops/worker_host.rs]: Why does this function take a tokio runtime? Could it not just called tokio_util::run_basic?

Comment (bartlomieju) [reply cli/ops/worker_host.rs]: WebWorkers will load modules using Tokio runtime before running the loop - so they need to create runtime manually beforehand

File: cli/ops/worker_host.rs [hunk 8]
```diff
@@ -136,31 +252,45 @@ struct WorkerArgs {
...
-
-    let mut receiver =
-      futures::executor::block_on(worker_handle.receiver.lock());
-    receiver.close();
-  };
-
+  let worker_handle =
+    state.workers.remove(&id).expect("No worker handle found");
+  worker_handle.terminate();
   Ok(JsonOp::Sync(json!({})))
```
Comment (ry) [inline cli/ops/worker_host.rs]: Nice

File: cli/tests/026_workers.ts [hunk 9]
```diff
@@ -1,26 +0,0 @@
...
-  tsWorker.postMessage("Hello World");
-};
-
-jsWorker.onerror = (e: Event): void => {
-  e.preventDefault();
-  console.log("called onerror in script");
-  jsWorker.postMessage("Hello World");
-};
-
-jsWorker.postMessage("Hello World");
```
Comment (ry) [inline cli/tests/026_workers.ts]: Does this test pass tho? Maybe we should just leave it even though it's duplicated in the unit tests. It's a nice little stand alone example.

File: cli/web_worker.rs [hunk 10]
```diff
@@ -77,19 +79,18 @@ mod tests {
       state,
     );
     worker.execute("bootstrapWorkerRuntime(\"TEST\")").unwrap();
-    worker.execute("runWorkerMessageLoop()").unwrap();
     worker
   }
-
+  // FIXME(bartlomieju)
+  #[ignore]
   #[test]
```
Comment (ry) [inline cli/web_worker.rs]: Disabled test which is also modified. Are these changes intentional?

File: cli/ops/worker_host.rs (hunk 7 above)
Comment (ry) [reply cli/ops/worker_host.rs]: Is the error you get something like "cannot call tokio::spawn outside of a tokio context" ? I wouldn't think that just creating the runtime is enough to solve that... Also if that's the error I wonder where it originates from if not during the event loop.

Comment (bartlomieju) [reply cli/ops/worker_host.rs]: > Is the error you get something like "cannot call tokio::spawn outside of a tokio context" ? I wouldn't think that just creating the runtime is enough to solve that... Also if that's the error I wonder where it originates from if not during the event loop.

I did not encounter such error here

File: cli/tests/026_workers.ts (hunk 9 above)
Comment (bartlomieju) [reply cli/tests/026_workers.ts]: Reverted

File: cli/web_worker.rs (hunk 10 above)
Comment (bartlomieju) [reply cli/web_worker.rs]: Fixed

File: cli/web_worker.rs (hunk 6 above)
Comment (bartlomieju) [reply cli/web_worker.rs]: Fixed

File: cli/ops/worker_host.rs (hunk 7 above)
Comment (ry) [reply cli/ops/worker_host.rs]: So if you don't pass the runtime through, what is the error you get? `rt` doesn't seem to be used in `run_worker_loop` except by `block_on`.

Comment (bartlomieju) [reply cli/ops/worker_host.rs]: I believe I did that because of `worker_is_ready` helper variable - got a big error when I tried to make `run_worker_loop` a closure that captures it. I'm pretty sure I'll refactor this in the next pass

**Comment by base698** [comment]
@bartlomieju When I do new Worker() in 0.33 it crashes saying reserved word?  How do you use the workers now?

Cheers

**Comment by ry** [comment]
@base698 here's an example https://github.com/denoland/deno/blob/v0.33.0/cli/tests/026_workers.ts

**Comment by base698** [comment]
@ry I'm creating a bundle with Deno.bundle and using that in webworker.  In 0.32 and prior it works.  Now it fails and crashes with encountered reserve word.

**Comment by bartlomieju** [comment]
@base698 seems like a bug, could you open a new issue and show some stack trace? It sounds like SystemJS and Workers don't like each other

**Comment by kitsonk** [comment]
(and provide an example of what you are bundling, I am not too sure it is limited to web workers)

**Comment by base698** [comment]
```worker.ts
function entryPoint(data) {
   console.log('received');
   postMessage("Hi 2 u");
}

if (self.name !== "workerFn") throw Error("Bad name");

onmessage = function(e) {
    entryPoint(e.data);
}

onerror = function(err) {
    console.error('on error', err);
    return false;
}
```

```01_worker.ts
var [diagnostics, emit] = await Deno.bundle('./worker.ts');
const encoder = new TextEncoder();
Deno.writeFileSync('./worker.js', encoder.encode(emit));

const workerFn = new Worker('./worker.js', { name: 'workerFn', type: "module" });
workerFn.onmessage = console.log;

workerFn.postMessage('from client');
```

### Deno 0.32
```
$ deno --allow-write 01_worker.ts
received
{ data: "Hi 2 u" }
```
### Deno 0.33
```
$ deno --allow-write 01_worker.ts
Compile file:///home/justin/workspaces/deno-test/tests/01_worker.ts
error: Uncaught Error: Uncaught TypeError: Cannot destructure property 'deps' of 'mMap.get(...)' as it is undefined.
► $deno$/workers.ts:169:17
    at poll ($deno$/workers.ts:169:17)
```

Unrelated, but `flags.rs` has v0.29 prepending.  I ran some tests in the test folder and it was trying to pull from std@v0.29.0 which was 404'ing.  If you have run deno when that was available and it was cached then it doesn't 404.

**Comment by base698** [comment]
@bartlomieju

**Comment by kitsonk** [comment]
It would have helped if this was a new issue as requested.

Also, not that we should throw, but is there a reason why you are bundling a single module? I think the problem us that when only one module is compiled, it doesn't provide a module name, and we are not predicting that properly, meaning we aren't initting the "bundle" properly.
